src/core/server/saved_objects/service/lib/repository.ts

```diff
@@ -738,12 +913,154 @@ export class SavedObjectsRepository {
       id,
       type,
       updated_at: time,
-      version: encodeHitVersion(response),
+      version: encodeHitVersion(updateResponse),
+      ...(this._registry.isMultiNamespace(type) && {
+        namespaces: updateResponse.get._source.namespaces,
+      }),
       references,
       attributes,
     };
   }
 
+  async addNamespaces(
+    type: string,
+    id: string,
+    namespaces: string[],
+    options: SavedObjectsAddNamespacesOptions = {}
+  ): Promise<{}> {
+    if (!this._allowedTypes.includes(type)) {
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    if (!this._registry.isMultiNamespace(type)) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(`${type} doesn't support namespaces`);
+    }
+
+    if (!Array.isArray(namespaces) || !namespaces.length) {
+      throw SavedObjectsErrorHelpers.createBadRequestError(
+        'namespaces must be a non-empty array of strings'
+      );
+    }
+
+    const { version, namespace, refresh = DEFAULT_REFRESH_SETTING } = options;
+
+    const rawId = this._serializer.generateRawId(undefined, type, id);
+    const preflightResult = await this.preflightCheckIncludesNamespace(type, id, namespace);
+    const existingNamespaces = getSavedObjectNamespaces(undefined, preflightResult);
+    // there should never be a case where a multi-namespace object does not have any existing namespaces
+    // however, it is a possibility if someone manually modifies the document in Elasticsearch
+    const time = this._getCurrentTime();
+
+    const doc = {
+      updated_at: time,
+      namespaces: existingNamespaces ? unique(existingNamespaces.concat(namespaces)) : namespaces,
+    };
+
+    const updateResponse = await this._writeToCluster('update', {
+      id: rawId,
+      index: this.getIndexForType(type),
+      ...getExpectedVersionProperties(version, preflightResult),
+      refresh,
+      ignore: [404],
+      body: {
+        doc,
+      },
+    });
+
+    if (updateResponse.status === 404) {
+      // see "404s from missing index" above
+      throw SavedObjectsErrorHelpers.createGenericNotFoundError(type, id);
+    }
+
+    return {};
+  }
+
+  async removeNamespaces(
```

Comment:
We don't need jsdoc comments for each of the parameters like the other functions (these are leftover from pre-ts days), but can you add a block comment to document the function? We should specifically highlight the "delete document if this is the only remaining namespace" behaviour. I see you added a doc comment to the SavedObjectsClient most API consumers will use that, but there are a few uses cases where plugins consume the repository directly.

Maybe we should call this method `deleteFromNamespaces` (and the other `addToNamespaces`)? I think that would make it clearer that this behaves like a delete where if the last namespace is removed we delete the document.

Reply:
Will do!

Sure, I can rename it to something that is clearer. Originally this was one method, `updateNamespaces`, but we split it out into two methods to simplify the authZ checks that we do in the Security plugin.

There isn't any such thing as "a namespace" per se, but I think intuitively `deleteFromNamespaces` and `addToNamespaces` makes more sense anyway. @legrego do you have any input before I rename these?

Reply:
Agree that namespaces aren’t really “things”, but I do like the suggested names. +1 from me

Reply:
OK, renamed methods in 418f1cd and added block comments in e2f5d65.

Reply:
Thanks! 🙇